### DISCUSSION TITLE
Uncaught (in promise) Component not found (V3)

### DISCUSSION THREAD

**Discussion by Muffinman** [livewire/livewire#7137] [7. Bugs (archived)] [answered]
Hi, I'm getting the following Livewire JS error when using a paginated component:

```
Uncaught (in promise) Component not found: uTeOY7srwQpnWa3hnp2U
```

Not sure if this is a V3 bug or something I'm doing wrong, but I'm leaning towards bug?

<details>
<summary>Version Info</summary>
<code>
Environment ..................................................................
Application Name ..................................................... Laravel
Laravel Version ...................................................... 10.28.0
PHP Version ............................................................ 8.2.7
Composer Version ....................................................... 2.5.8
Environment ............................................................ local
Debug Mode ........................................................... ENABLED
URL .................................................... super secret stuff
Maintenance Mode ......................................................... OFF

Cache ........................................................................
Config ............................................................ NOT CACHED
Events ............................................................ NOT CACHED
Routes ............................................................ NOT CACHED
Views ................................................................. CACHED

Drivers ......................................................................
Broadcasting ............................................................. log
Cache .................................................................. redis
Database ............................................................... mysql
Logs .......................................................... stack / single
Mail .................................................................... smtp
Queue .................................................................. redis
Scout .............................................................. typesense
Session ................................................................. file

Livewire .....................................................................
Livewire ............................................................. v3.0.10
</code>

</details>

I have two Livewire components, with one included indirectly via the other.

`product-list.blade.php`
```blade
<div>
    <label for="in_stock" class="cursor-pointer">
        <input type="checkbox" wire:model.live="in_stock" name="in_stock" id="in_stock" />
        Only show items in stock
    </label>

    {{ $products->links() }}

    <div class="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 xl:grid-cols-5 gap-10 gap-y-12 mt-10 transition-all duration-300 mb-10" wire:loading.class="opacity-0 max-h-0">
        @foreach ($products as $product)
            <x-product-teaser :product="$product" :category="$category" />
        @endforeach
    </div>

    <div wire:loading.flex class="flex items-center justify-center">
        <span class="block loading loading-spinner text-primary w-8 h-8 my-8"></span>
    </div>

    {{ $products->links() }}
</div>
```

The `x-product-teaser` blade component simply renders the product details but also includes another Livewire component. For the purposes of this demo I have simplified by ONLY including a livewire component:

`product-teaser.blade.php`
```blade
<div>
    <livewire:product-single-add wire:key="product-single-add-{{ $product->id }}" :product="$product" :instance_id="$product->instance_id" />
</div>
```

`product-single-add.blade.php`
```
<div>
    Test here
</div>
```

So the heirarchy is:

```
  -> `product-list.blade.php` (Livewire, paginated) 
      -> `product-teaser.blade.php` (Blade)
          -> `product-single-add.blade.php` (Livewire)
```

![Screenshot 2023-10-23 at 15 12 19](https://github.com/livewire/livewire/assets/1319568/61c7af03-82b8-4232-97d8-9cd5e357544b)

Whenever I interact with the pagination or the filter tickbox it works the first time (although the console error is produced). Subsequent interactions no longer work and the Wire IDs mentioned in the log messages don't match the DOM IDs.

The following things **DO** to work, so I think I must have stumbled on a strange combination.

- Changing `wire:model.live="in_stock"` to `wire:model="in_stock"`. No errors now. Pagination & filtering works properly (albeit not instant any more).
- Commenting out `<livewire:product-single-add />` in final included template
- Moving ` <livewire:product-single-add />` into the top template directly. No idea why this works as the HTML produced is virtually identical other than the extra wrapping div.

Any ideas?

[Reproduction repo](https://github.com/Muffinman/livewire-bug-report)

**Comment by lrljoe** [comment]
Two key questions here:

**1. What's the purpose of the child Livewire component "product-single-add"?**  
Looking at your screenshot, I see no reason for it to be a Livewire component in of itself.  It could be a blade component easily I'd imagine.

**2. Why are you using a blade, just to add a nested component in?**  
I see no benefit at all in doing that, you're just rendering a blade needlessly, and having to track it, look at you should be keying nested components in the docs below:
[See Docs](https://livewire.laravel.com/docs/nesting#rendering-children-in-a-loop)

Not a bug as far as I can see, just a misunderstanding of how to use nested components.

**Comment by Muffinman** [comment]
Thanks for the reply.

The templates are far more complex than I posted above, but I've stripped most things out for brevity, the problem still exists in the simplified format above.

The single-add component is livewire because it has live quantity stock checks and triggers events in other parts of the template. Wasn't relevant to the problem above but it is needed.

The teaser blade template is used because we re-use this same display in several different places throughout the site.

I have correctly keyed the child component in the example above, still looks like a bug to me!

I will try to create a reproduction repo and share.

**Reply by lrljoe** [reply]
You should use :key= not wire:key= on a Livewire component

unless I'm having a moment!

**Reply by Muffinman** [reply]
I believe either are valid, depending on if you're passing a variable or a string:

https://livewire.laravel.com/docs/components#adding-wirekey-to-foreach-loops

**Reply by lrljoe** [reply]
Per the docs, use :key to pass to nested Livewire components, wire:key is for inline elements or nested blade components (again I could be going mad here)

Always consider why you are nesting livewire components, there is rarely a good reason to!

**Comment by Muffinman** [comment]
I have recreated the issue in a simple repository:

https://github.com/Muffinman/livewire-bug-report

**Reply by Muffinman** [reply]
EDIT: I have discovered that the problem is definitely linked to the `key` on the child component.

If I replace it the fixed string with something random then everything works as expected.
```blade
<livewire:product-single-add wire:key="{{ Str::uuid() }}" :product="$product" />
```

**Comment by Muffinman** [comment, marked as answer]
OK I found the issue, and it was my fault for not reading the docs thoroughly enough.

The `wire:key` has to be on the *root* element **DIRECTLY** inside the foreach loop. From the docs:

> When looping through data in a Livewire template using @foreach, you must add a unique wire:key attribute to the root element rendered by the loop.

https://livewire.laravel.com/docs/components#adding-wirekey-to-foreach-loops

What is not clear from the docs is that if you have a CSS grid it seems you need to do something like:

```blade
<div class="grid grid-cols-12">
    @foreach ($items as $item)
        <div class="col-span-3" wire:key="title-{{ $item->id }}">{{ $item->title }}</div>
        <div class="col-span-3" wire:key="price-{{ $item->id }}">{{ $item->price }}</div>
        <div class="col-span-3" wire:key="something-else-{{ $item->id }}">{{ $item->something_else }}</div>
    @endforeach
</div>
```

Adding unique keys to each element.

**Reply by QuentinGab** [reply]
Pro tips:

You can simply use the CSS: [property](https://developer.mozilla.org/en-US/docs/Web/CSS/display#display_contents) `display: contents` on a div like this

```blade
<div class="grid grid-cols-12">
    @foreach ($items as $item)
       <div class="contents" wire:key="items.{{ $item->id }}">
            <div class="col-span-3">{{ $item->title }}</div>
            <div class="col-span-3">{{ $item->price }}</div>
            <div class="col-span-3">{{ $item->something_else }}</div>
      </div>
    @endforeach
</div>
```

**Comment by wonderful-jason-rudland** [comment]
If it helps anyone I had exactly the same error. I followed the answer in this thread, but to no avail. Eventually I discovered I had more than a single element in my view. i.e.
```
<div>
   [all my component HTML]
</div>
<div>This div was mistakenly moved outside the main div above</div>
```
The console error happened when I dispatched an event (any event), like this:
```
$this->dispatch('post-creted')
```
It took me forever to figure it out as the error seemed totally unrelated to what was actually wrong.

**Comment by rizkyilhampra** [comment]
Just add another workaround for this issue. In my case, it was caused by a missing blade component. For example, I have a livewire/volt component like this:
```blade
<div>
    <x-card>
    ...
    </x-card>
</div>
```
After spending some hours finding the cause of this problem, I realized that I had forgotten to add the card component in `resources/views/components`. I tried adding `wire:key` in the foreach loops, but it didn't help because that wasn't the main issue in my case.

**Comment by ManuDoni** [comment]
I had the same error but none of the other solutions worked for me.
In my case, I was setting `wire:key` on all loops and `:key` on nested livewire components (all of them with prefixes) but to make it work I had to add a filter value on the key.

The filter value `$matchesFilter` was filtering the rows of a list and inside every list item i had a livewire nested component.
Before:
```blade
<livewire:place-bet-button :key="'bet_button_'.$match->id" />
```
After:
```blade
<livewire:place-bet-button :key="'bet_button_'.$match->id.$matchesFilter" />
```

**Comment by martinsnajdr** [comment]
I get the same error when I use Media Library Pro Livewire component and the components is just a part of multistep form. When the step changes and the media library component is not present in next step anymore, I get this error. 

Tried to add keys everywhere.

Maybe it's just my basic mistake, but Livewire can't handle when the nested Livewire component (inside a fullpage one) is removed (because there is if condition)?
